### ISSUE TITLE
Slow rendering with specific cells

### ISSUE THREAD

**Issue by gunterkoenigsmann** [wxMaxima-developers/wxmaxima#1419]
wxMaxima has grown to be real fast. But the file that triggered bug #1415 causes the displaying routines to be rather slow on my Linux/GTK3 computer.

I cannot see spurious recalcs => It is the draw routines this time, somehow. Might it be that we draw stuff more than once per refresh?

**Comment by KubaO** [comment]
There has been a performance regression recently, and it seems it has been major. I'll bisect and see when it happened.

**Comment by KubaO** [comment]
Initial bisecting:
2020-09-02 - 3m50s
2020-10-11 - 4m30s
2020-10-23 - >7m20s

**Comment by KubaO** [comment]
The "Cell list recalc fixes" 42c90bea made benchmarks run at least 2.5x slower, so whatever was the reasoning that led to that change was misguided: it really makes things much worse. I bet that if you go right before the "Cell list recalc fixes", you'll experience normal performance again. That's certainly the case for me.

> Now we recalculate all fast-to-recalc-cells anew every time their GroupCell recalculates itself. But on the other hand recalculation now is simpler.

Nothing is fast to recalculate when you got a million of it. To evaluate performance, benchmarks must do actual work, hundreds of thousands of times, and the `hs` benchmark is very good for that. The mentioned commit causes multiple recalculation of the cell list anchored in the group cell, and when a single recalculation takes 4 minutes, doing it multiple times makes the result easy to foresee.

**Comment by KubaO** [comment]
Subsequent changes made it even worse - at the moment the performance is at least 4x worse than it was right before 42c90be.

**Comment by KubaO** [comment]
Effectively, the draw list removal undoes all that damage, and makes the benchmarks almost 10x faster, so that will be my focus.

**Comment by gunterkoenigsmann** [comment]
If you can do anything about this that  would be great: commit 42c90be tried to resolve the problem that linebreaks were broken if parenthesis, fractions, square roots or any other objects that can be broken into lines were nested.

**Comment by KubaO** [comment]
Yep, and that commit definitely did fix that problem. Since the "penalty" was only a 2x lenghtening of the benchmark, it's not that big of an issue given that visual glitches got fixed. But it'd still be interesting if you could build from before that commit and see if the performance of this problematic snippet is better. That would be some sort of a corner case since it's a very simple document to lay out, and very short, so I'm surprised that it would feel slow at all. Normally such small documents are too small to notice any performance regressions (at least for me).

**Comment by gunterkoenigsmann** [comment]
On my computer [Ubuntu 20.10] I don't see a notable difference between before my merge and afterwards for this specific problem.

...and as I have changed only the recalculating code and the only thing that is slow in this case is the displaying of equation (which should not cause recalculations if the window size and resolution isn't changed) I would have expected that.

What I see is a difference of the recalc speed for big files with long nested square roots and parenthesis. But as my commit caused these objects to be recalculated twice (once as 2D object in order to find out if they need to be broken into lines and once as the 1D object they are converted to since they need to be broken into lines) that might be OK.

Now I am looking forward to the draw list removal: You often surprise me how fast complicated tasks can get if done right.

**Comment by gunterkoenigsmann** [comment]
Don't know if that is a new or if it is this issue. But the attached file [renamed from .wxmx to .txt in order to convince github that it is save attaching it] is suspiciously slow, too.
[binm19.txt](https://github.com/wxMaxima-developers/wxmaxima/files/5713723/binm19.txt)

**Comment by gunterkoenigsmann** [comment]
Resolved by now.
